design/external-references.md

```diff
@@ -0,0 +1,344 @@
+  writeString(key: string, value: string, options?: SerializationOptions): void
+  writeObject(key: string, obj: ISerializable, options?: SerializationOptions): void;
+}
+
+interface SerializationOptions {
+  description?: string;
+}
+```
+
+The serialization context allows the object serialize itself through key/value strings via calls to `writeString(key, value)`. If the object encapsulates another serializable object, it can use `writeObject(key, obj)`, which will result in a subsequent call to the sub-object's `serialize` method with an appropriate context.
+
+### Deserialization
+
+To support deserialization, classes must also include a public static `deserializeXxx` method which reads the object from a deserialization context and returns an object that implements the resource type interface:
+
+```ts
+class MyResource extends Construct implements ISerializable {
+  static deserializeMyResource(ctx: IDeserializationContext): IMyResource;
+}
+```
+
+The deserialization context is an object that implements the following interface:
+
+```ts
+interface IDeserializationContext {
+  scope: Construct;
+  id: string;
+  readString(key: string): string;
+  readObject(key: string): IDeserializationContext;
+}
+```
+
+`readString(key)` can be used to read values stored by `writeString`. and `readObject(key)` returns a deserialization context for composite deserialization written via `writeObject`.
+
+Since `deserializeXxx` will normally need to create new construct objects, the deserialization context should supply a consistent `scope` and `id` which can be used to instantiate a construct object that represents this object.
+
+Implementers of `deserializeXxx` should check if a construct with `id` already exists within `scope` and return it instead of instantiating and new object. This will satisfy REQ6.
+
+The reason we indicate the type in the method name is because static methods in JavaScript are inherited, so we can differentiate between `ExtraBucket.deserializeExtraBucket` and `ExtraBucket.deserizlizeBucket`.
+
+---
+
+Here's an example that demonstrates how serialization and deserialization can be implemented for an application load balancer (ALB), which is a composite resource. As you can see, ALBs encapsulate a security group, which is serialized and deserialized together with the ALB itself.
+
+```ts
+class ApplicationLoadBalancer {
+  serialize(ctx: ISerializationContext): void {
+    ctx.writeString('LoadBalancerArn', this.loadBalancerArn);
+    ctx.writeObject('SecurityGroup', this.securityGroup);
+  }
+
+  static deserializeApplicationLoadBalancer(ctx: IDeserializationContext): IApplicationLoadBalancer {
+    const exists = ctx.scope.findChild(ctx.id);
+    if (exists) {
+      return exists;
+    }
+
+    return new ImportedApplicationLoadBalancer(ctx.scope, ctx.id, {
+      loadBalancerArn: ctx.readString('LoadBalancerArn'),
+      securityGroup: ec2.SecurityGroup.deserializeSecurityGroup(ctx.readObject('SecurityGroup'))
```

Comment:
How does `readObject()` know how to change `scope` and `id` for the inner object? 

It has to, otherwise the SG and the LB will be created in the same scope and with the same ID. But I just don't see the implementation of `readObject`, how will it know what to change `scope` and `id` to?

Reply:
Let me know if the below examples clarify this a little

Reply:
Not sure which example you're referring to.

Reply:
Basically readObject creates a new deserialization context that’s “derived” from the parent context that contains the scope (probably will always be the stack or some global scope maintained by the deserializer) and a unique id that’s derived from the fully qualified export name. Then, the deserizlize method (in this case “deserializeSecurityGroup”) will implement the same pattern depicted here to conditionally create a new instance.